### PR TITLE
feat: Horizontal filter bar states

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY
This PR implements the horizontal container of the FilterBar and splits the component into two, Vertical and Horizontal.

### AFTER

https://user-images.githubusercontent.com/60598000/200869208-44cfeda9-8bc5-455f-a054-f553aeaf6c00.mp4

### TESTING INSTRUCTIONS
1. Set HORIZONTAL_FILTER_BAR feature flag to True
2. Go to a Dashboard and open the left sidebar
3. Click on the gear icon to change the FilterBar orientation to horizontal

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags: HORIZONTAL_FILTER_BAR
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx:364]: Can you extract `nativeFiltersEnabled && !editMode` to a const called `showFilterBar` to be used in both places?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/index.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/index.tsx:41]: Can we rename it to `FilterBarOrientation`?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/index.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/ActionButtons/index.tsx:44]: Totally optional, but for these cases, I like to use [Emotion's Composition](https://emotion.sh/docs/composition).

You can write the individual styles:

```
const containerStyle = (theme: SupersetTheme) => css`
  display: flex;

  && > .filter-clear-all-button {
    color: ${theme.colors.grayscale.base};
    margin-left: 0;
    &:hover {
      color: ${theme.colors.primary.dark1};
    }

    &[disabled],
    &[disabled]:hover {
      color: ${theme.colors.grayscale.light1};
    }
  }
`;

const verticalStyle = (theme: SupersetTheme, width: number) => css`
  flex-direction: column;
  align-items: center;
  pointer-events: none;
  position: fixed;
  z-index: 100;

  // filter bar width minus 1px for border
  width: ${width - 1}px;
  bottom: 0;

  padding: ${theme.gridUnit * 4}px;
  padding-top: ${theme.gridUnit * 6}px;

  background: linear-gradient(
    ${rgba(theme.colors.grayscale.light5, 0)},
    ${theme.colors.grayscale.light5} ${theme.opacity.mediumLight}
  );

  & > button {
    pointer-events: auto;
  }

  & > .filter-apply-button {
    margin-bottom: ${theme.gridUnit * 3}px;
  }
`;

const horizontalStyle = (theme: SupersetTheme) => css`
  margin: 0 ${theme.gridUnit * 2}px;
  && > .filter-clear-all-button {
    text-transform: capitalize;
    font-weight: ${theme.typography.weights.normal};
  }
  & > .filter-apply-button {
    &[disabled],
    &[disabled]:hover {
      color: ${theme.colors.grayscale.light1};
      background: ${theme.colors.grayscale.light3};
    }
  }
`;
```

And apply them conditionally:

```
return (
    <div
      data-test="filterbar-action-buttons"
      css={(theme: SupersetTheme) => [
        containerStyle(theme),
        isVertical ? verticalStyle(theme, width) : horizontalStyle(theme),
      ]}
    >
      ...
    </div>
  );
};
```

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Header/index.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Header/index.tsx:65]:
```suggestion
const Wrapper = styled.div`
  ${({ theme }) => `
    padding: ${theme.gridUnit}px ${theme.gridUnit * 2}px;

    .ant-dropdown-trigger span {
      padding-right: ${theme.gridUnit * 2}px;
    }
  `}
`;
```

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Horizontal.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Horizontal.tsx:79]:
```suggestion
const HorizontalBar = styled.div`
  ${({ theme }) => `
    padding: ${theme.gridUnit * 2}px ${theme.gridUnit * 2}px;
    background: ${theme.colors.grayscale.light5};
    box-shadow: inset 0px -2px 2px -1px ${theme.colors.grayscale.light2};
  `}
`;

const HorizontalBarContent = styled.div`
  ${({ theme }) => `
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: center;
    justify-content: flex-start;
    padding: 0 ${theme.gridUnit * 2}px;

    .loading {
      margin: ${theme.gridUnit * 2}px auto ${theme.gridUnit * 2}px;
      padding: 0;
    }
  `}
`;

const FilterBarEmptyStateContainer = styled.div`
  ${({ theme }) => `
    margin: 0 ${theme.gridUnit * 2}px;
    font-weight: ${theme.typography.weights.bold};
  `}
`;

const FiltersLinkContainer = styled.div<{ hasFilters: boolean }>`
  ${({ theme, hasFilters }) => `
    padding: 0 ${theme.gridUnit * 2}px;
    border-right: ${
      hasFilters ? `1px solid ${theme.colors.grayscale.light2}` : 0
    };

    button {
      display: flex;
      align-items: center;
      text-transform: capitalize;
      font-weight: ${theme.typography.weights.normal};
      color: ${theme.colors.primary.base};
      > .anticon + span, > .anticon {
          margin-right: 0;
          margin-left: 0;
        }
    }
  `}
`;
```

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/HorizontalFilterBar.test.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/HorizontalFilterBar.test.tsx:25]: [Avoid nesting when you're testing](https://github.com/apache/superset/wiki/Testing-Guidelines-and-Best-Practices#avoid-nesting-when-youre-testing) 😉

**Review by michael-s-molina — COMMENTED** [review]
Thanks for the PR @geido. I still have some files to review but I'll leave some comments already in case you want to work on them. Awesome work!

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx:67]: Maybe promote `theme` to the top level to avoid repeating code?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx:82]: Same here

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx:98]: Same here

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Vertical.tsx:164]: This is functionally equivalent without the warning.

```suggestion
  const onScroll = useMemo(
    () =>
      throttle(() => {
        clearTimeout(timeout.current);
        setIsScrolling(true);
        timeout.current = setTimeout(() => {
          setIsScrolling(false);
        }, 300);
      }, 200),
    [],
  );
```

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:256]: We have many missing dependencies here. I suggest not using `useMemo` if we can't handle the dependencies because it will be using stale values.

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/utils.ts
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/utils.ts:31]: Can we leave only utility functions in this file and move type definitions to a types file?

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx
Comment (codyml) [inline superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx:369]: Getting a missing dependency warning about `directPathToChild`.

**Review by codyml — COMMENTED** [review]
Tested locally and it works!  Left some comments.

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Horizontal.tsx
Comment (geido) [reply superset-frontend/src/dashboard/components/nativeFilters/FilterBar/Horizontal.tsx]: > I think according to the designs, distance between the gear and this message should be 18px.

Increased to 16px which seems to be the most common spacing value in the horizontal filter bar. I think it looks balanced this way

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/HorizontalFilterBar.test.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/HorizontalFilterBar.test.tsx:100]:
```suggestion
test('should not render Add/Edit Filters', async () => {
```

**Review by michael-s-molina — APPROVED** [review]
Code LGTM. Thanks for all the hard work and for addressing the comments @geido! I suggest waiting for @jinghua-qa's approval before merging it.

**Comment by jinghua-qa** [comment]
Regression test LGTM

**Pull request merged**
